core/src/main/java/io/grpc/internal/AbstractClientStream2.java

```diff
@@ -248,12 +248,6 @@ protected void inboundDataReceived(ReadableBuffer frame) {
           log.log(Level.INFO, "Received data on closed stream");
           return;
         }
-        if (!headersReceived) {
```

Comment:
`headersReceived` is now unused. Either delete it, or turn this into an `IllegalStateException` (so that we get a clear error that a class is mis-implemented).